Added Amazon S3 RrdBackend support

pmelici opened the pull request:
Hi there -- Thanks for the great project! I had a need for Amazon S3 backend support, so I went ahead and implemented it. I also added the idea of RrdBackendCompression, which is currently only implemented for RrdS3Backend, but the other custom backends could take advantage of this as well if you want them to. 

There is a unit test for the RrdS3Backend. If you want to run it, you will need an S3 account (you can get a free one) and then set system env variables to point to your account. Details on this will be shown by the test if you attempt to run it without an S3 account available.

I changed the RrdDbTest a bit just so I could re-use the code, but did not make any functional changes. I also made RrdDb implement Closable so that it can be used with the try-with-resources syntax in java.

Let me know if you have any questions, concerns, etc. Hoping you will decide to merge this and put out a new release at some point soon so I can go back to depending on your official project in the bigger proprietary project I'm working on.

Cheers,
Paul Melici

fbacchella commented:
I'm ok to merge a S3 backend but I have issue with yours.
First don't include irrelevant patch, like the .gitignore one., same thing for RrdDb being closable.
Using environment variable is not a java compliant way. Please use a http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/SystemPropertiesCredentialsProvider.html and http://www.mojohaus.org/properties-maven-plugin/.
tryRegisterAndSetAsDefaultFactory is not a bad method, but should not be used only in one factory. Please remove it.

pmelici commented:
1. I reverted the change to RrdDb making it Closable and the .gitignore changes.
2. I have removed tryRegisterAndSetAsDefaultFactory. But just to shed a bit of light on why I created it: simply calling registerAndSetAsDefaultFactory like the other rrd backend factories makes it impossible to create multiple instances of the factory with different settings. So, let’s say you have an application that needs to write to multiple MongoDB databases or multiple S3 buckets – that’s not possible currently. Moreover, the RrdBackendS3Test fails (because it create multiple instances with different settings for testing purposes) if just the normal registerAndSetAsDefaultFactory is used. 
So, for now, I have removed the tryRegisterAndSetAsDefaultFactory method as you requested and RrdS3BackendFactory now calls the normal registerAndSetAsDefaultFactory(), but suppresses IllegalStateException / IllegalArgumentException so that it’s possible to create multiple instances with different configurations. This is obviously not an ideal solution. If there is a better approach that you would prefer that allows for multiple instances of factories to be created with different settings, please let me know.
3. I agree that environmental variables are not frequently used in Java, but none-the-less this is the default and recommended means of configuration as per http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html. I may not have mentioned it before, but any of the configuration methods supported by AmazonS3ClientBuilder.standard() will work both for production and unit testing purposes, in addition to the environmental variables. For example, you could create a ~/.aws/credentials file instead of setting AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.
But in terms of what you suggested regarding using maven to read a properties file and pass system variables to Java: Yes, there is a SystemPropertiesCredentialsProvider that will get the credentials into the S3 client. But the S3 client needs both credentials and a region. However there is not system properties configuration provider for the region in the AWS SDK. The available ones are listed under Direct Known Subclasses in the Javadoc here: http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/regions/AwsRegionProvider.html -- none support java system properties.
So even if the SystemPropertiesCredentialsProvider is used for credentials, then there is still the matter of the region configuration – which would either still need to be set as an environmental variable (which seems inconsistent to get the credentials from java system properties and the region from environmental variable), or a custom AwsRegionProvider could be implemented to read from java system properties (but writing a custom AwsRegionProvider feels a bit out of place within rrd4j).
My thoughts are we should stick with the default / standard configuration, which supports environmental variables as well as system properties for the credentials config, and supports environmental variables (and other means) for the region config.

fbacchella commented:
I don't think I will merge your code.

It's only specific code and you don't need it to be merged in main tree.

Registering external external backend is as simple as:

        RrdBackendFactory f1 = new MyBackendFactory(...);
        RrdBackendFactory f2 = new MyBackendFactory(...);

        RrdDb db1 = new RrdDb(..., f1);
        RrdDb db2 = new RrdDb(..., f2);

If each instance of MyBackendFactory get a different name, and they are registered, accessing them can be easy:

    RrdBackendFactory f = new MyBackendFactory(...);
    RrdBackendFactory.register(f);
    ....
    RrdDb db = new RrdDb(..., RrdBackendFactory.register.getFactory("F1NAME"));

It this fails, please notify me, and I will correct that.

I merged the Closeable trick. It's a good idea and I should have done that a long time ago.

pmelici commented:
Yes, you're right, a custom backend outside of the main code base should work just fine for me. 

But is the issue with merging the code just the environmental variables needed to run the unit tests? If so, there is the option to make it work with a .properties file and I'll implement that if you like.

Thanks

fbacchella commented:
No it's more about adding too much external and very specific dependencies in the project. But an subproject under the rrd4j "organisation" is welcome.
That's should be the way to go for mongodb and berkleydb, but they are already in project. I will perhaps remove them in a possible version 4 of rrd4j.

fbacchella commented:
I changed the way backend are manage. They can be identified using URI, see my commit a4a91c23a905a34038a3b29b4b56c1ab4a532782.

You just don't register a backend factory as a default, but ensure that instances of the same kind can identify different URI and put the in the active list.
